content/codes/components/enigma.ts

```diff
@@ -6,31 +6,598 @@
+	return ch.charCodeAt(0) - charA;
+}
+
+function getLast(a:string[]):string
+{
+  return a.length ? a[a.length - 1] : "";
+}
+
+interface Substitution {
+  [key: string]: string
+}
+
+class Plugboard {
+  private count:number = 0;
+  private connections:Substitution = {};
+
+  getConnections()
+  {
+    return this.connections;
+  }
+
+	connect(letter1:string, letter2:string)
+	{
+		if (letter1 === letter2)
+			return;
+		else if (this.count == 10)
+			return;
+		this.count++;
+		this.connections[letter1] = letter2;
+		this.connections[letter2] = letter1;
+	}
+	
+	disconnect(letter1:string)
+	{
+		var conn = this.connections[letter1];
+		if (!conn)
+			return;
+		this.count--;
+		delete this.connections[conn];
+		delete this.connections[letter1];
+	}
+	
+	encode(ch:string):string
+	{
+		var conn = this.connections[ch];
+		return conn ? conn : ch;
+	}
+		
+	toString():string
+	{  
+		var str = [...letters].map(l => this.connections[l] ? this.connections[l] : l).join("");
+		return "Plugboard\n\tABCDEFGHIJKLMNOPQRSTUVWXYZ\n\t" + str;
+	}
+}
+
+class Reflector
+{
+  private forward:Substitution = {};
+
+  constructor(reflections:string) 
+  {
+    for (var i = 0; i < letters.length; i++) 
+      this.forward[letters[i]] = reflections[i];
+  }
+
+  getConnections()
+  {
+    return this.forward;
+  }
+
+  reflect(letter:string):string
+  {
+    var index = charToIndex(this.forward[letter]);
+    return letters[index];
+  }
+		
+	toString():string
+	{  
+		var str = [...letters].map(l => this.forward[l]).join("");
+		return "Reflector\n\tABCDEFGHIJKLMNOPQRSTUVWXYZ\n\t" + str;
+	}
+}
+
+class Rotor
+{
+  private readonly index:number;
+  // countdown steps until we need to turnover the adjacent rotor
+  stepsToTurnover:number = 26;
+  private offset:number = 0;
+  private notch:string;
+  private initialPosition:string;
+  private ringSetting:string;
+  private forward:Substitution = {};
+  private reverse:Substitution = {};
+
+  constructor(index:number, wiringMap:string, notch:string, initialPosition:string, ringSetting:string)
+  {
+    this.index = index;
+
+    // build up the forward and reverse substitution maps
+    for (let i = 0; i < letters.length; i++) 
+    {
+      this.forward[letters[i]] = wiringMap[i];
+      this.reverse[wiringMap[i]] = letters[i];
+    }
+    
+    this.initialPosition = initialPosition;
+    this.ringSetting = ringSetting;
+    this.notch = notch;
+    this.stepsToTurnover = charToIndex(notch) + 1;
+    // TODO inelegant to just repeatedly step...
+    for (let i = 0; i < charToIndex(initialPosition); i++)
+      this.step();
+    for (let i = 0; i < 26 - charToIndex(ringSetting); i++) 
+    {
+      this.updateWires();
+      this.offset = ++this.offset % letters.length;
+    }
+  }
+
+  getOffset()
+  {
+    return this.offset;
+  }
+
+  getNotch()
+  {
+    return this.notch;
+  }
+
+  getForwardConnections()
+  {
+    return this.forward;
+  }
+
+  getReverseConnections()
+  {
+    return this.reverse;
+  }
```

Comment:
I think we can return all these getters, and instead just make the properties public